lib/Table.js

```diff
@@ -11,154 +11,299 @@ function Table(name, schema, options, base) {
   this.options = options || {};
   this.base = base;
 
-  if(this.options.create === undefined || this.options.create === null) {
+  if (this.options.create === undefined || this.options.create === null) {
     this.options.create = true;
   }
 }
 
+var compareIndexes = function compareIndexes(local, remote) {
+  var i;
+  var j;
+  var indexes = {
+    delete: [],
+    create: [],
+    both: []
+  };
+  var localTableReq = local;
+  var remoteTableReq = remote;
+
+  var localIndexes = localTableReq.GlobalSecondaryIndexes;
+  var remoteIndexes = remoteTableReq.GlobalSecondaryIndexes;
+
+  debug('compareIndexes');
+  // let's see what remote indexes we need to sync or create
+  for (i = 0; i < localIndexes.length; i++) {
+    var remoteIndexFound = false;
+    for (j = 0; j < remoteIndexes.length; j++) {
+      if (remoteIndexes[j].IndexName === localIndexes[i].IndexName) {
+        // let's see if the core data matches. if it doesn't,
+        // we may need to delete the remote GSI and rebuild.
+        var localIndex = _.pick(localIndexes[i], 'IndexName', 'KeySchema', 'Projection', 'ProvisionedThroughput');
+        var remoteIndex = _.pick(remoteIndexes[j], 'IndexName', 'KeySchema', 'Projection', 'ProvisionedThroughput');
+        if (remoteIndex.hasOwnProperty('ProvisionedThroughput')) {
+          delete remoteIndex.ProvisionedThroughput.NumberOfDecreasesToday;
+        }
+
+        if (!_.isEqual(remoteIndex, localIndex)) {
+          indexes.both.push(localIndex);
+          remoteIndexFound = true;
+        } else {
+          remoteIndexFound = true;
+        }
+      }
+    }
+    if (!remoteIndexFound) {
+      indexes.create.push(localIndexes[i]);
+    }
+  }
+  // now let's see what remote indexes exist that shouldn't exist
+  for (j = 0; j < remoteIndexes.length; j++) {
+    var localExists = false;
+    for (i = 0; i < localIndexes.length; i++) {
+      if (remoteIndexes[j].IndexName === localIndexes[i].IndexName) {
+        localExists = true;
+      }
+    }
+    if (!localExists) {
+      indexes.delete.push(remoteIndexes[j]);
+    }
+  }
+
+  return indexes;
+};
+
+Table.prototype.deleteIndex = function deleteIndex(indexName) {
+  debug('deleting index %s', indexName);
+  var deferred = Q.defer();
+  var table = this;
+  table.active = false;
+  var ddb = this.base.ddb();
+  var params = {
+    TableName: table.name,
+    GlobalSecondaryIndexUpdates: [
+      {
+        Delete: {
+          IndexName: indexName
+        }
+      }
+    ]
+  };
+  ddb.updateTable(params, function (err, data) {
+    debug('deleteIndex handler running');
+    if (err) {
+      deferred.reject(err);
+    }
+    else {
+      setTimeout(function () {
+        table.waitForActive();
```

Comment:
You need to wait for the `waitForActive` promise to return before resolving the current `deleteIndex` promise.

```js
table.waitForActive().then(function () {
  deferred.resolve(data);
});
```